manual/src/tutorials/effectexamples.etex

```diff
@@ -0,0 +1,666 @@
+performing the effect.
+
+Another point to note is that the catch-all case ``"| _ -> None"'' is necessary
+when handling effects. This case may be intuitively read as ``forward the
+unhandled effects to the outer handler''.
+
+\section{s:effects-concurrency}{Concurrency}
+
+The expressive power of effect handlers comes from the delimited continuation.
+While the previous example immediately resumed the computation, the computation
+may be resumed later, running some other computation in the interim. Let us
+extend the previous example and implement message-passing concurrency between
+two concurrent computations using the "Xchg" effect. We call these concurrent
+computations \textit{tasks}.
+
+A task may either be in suspended state or is completed. We represent the task
+status as follows:
+
+\begin{caml_example*}{verbatim}
+type 'a status =
+  Complete of 'a
+| Suspended of {msg: int; cont: (int, 'a status) continuation}
+\end{caml_example*}
+
+The task may either have been "Complete" with the result or is "Suspended" with
+the message "msg" to send and the continuation "cont". The type "(int, 'a
+status) continuation" says that the suspended computation expects an "int"
+value to resume and returns a "'a status" value when resumed.
+
+Next, we define a "step" function that executes one step of computation until
+it completes or suspends:
+
+\begin{caml_example*}{verbatim}
+let step f () =
+  match_with f ()
+  { retc = (fun v -> Complete v);
+    exnc = raise;
+    effc = fun (type a) (eff: a t) ->
+      match eff with
+      | Xchg msg -> Some (fun (cont: (a, _) continuation) ->
+          Suspended {msg; cont})
+      | _ -> None }
+\end{caml_example*}
+
+Like "try_with", "match_with" also introduces an effect handler. Unlike
+"try_with", where only the effect cases "effc" is provided, "match_with"
+expects the handlers for the value ("retc") and exceptional ("exnc") return
+cases.
+
+\begin{itemize}
+  \item Case "retc": If the computation returns with a value "v", we return
+    "Complete v".
+  \item Case "exnc": If the computation raises an exception, then the handler
+    raises the same exception.
+  \item Case "effc": If the computation performs the effect "Xchg msg" with the
+    continuation "cont", then we return "Suspended{msg;cont}".
+\end{itemize}
+
+We can now write a simple scheduler that runs a pair of tasks to completion:
+
+\begin{caml_example*}{verbatim}
+let rec run_both a b =
+  match a (), b () with
+  | Complete va, Complete vb -> (va, vb)
+  | Suspended {msg = m1; cont = k1},
+    Suspended {msg = m2; cont = k2} ->
+      run_both (fun () -> continue k1 m2)
+               (fun () -> continue k2 m1)
+  | _ -> failwith "Improper synchronization"
+\end{caml_example*}
+
+Both of the tasks may run to completion, or both may offer to exchange a
+message. In the latter case, we resume the computations with the value from the
+other. We consider the other cases to be improperly synchronized programs.
```

Comment:
Instead of "the other cases", I would suggest clarifying: "the situation where one computation offers an exchange while the other computation terminates is regarded as a programmer error, and causes the handler to raise an exception".